### ISSUE TITLE
add_membership affects previous options in gen_udp:open/2

### ISSUE THREAD

**Issue by door** [erlang/otp#6476]
**Describe the bug**
Option `binary` is ignored when placed before `add_membership` in gen_udp:open/2 call

**To Reproduce**
```
#!/usr/bin/env escript

-mode(compile).

main(_) ->
  {ok, R} = gen_udp:open(4321, [binary, {add_membership,{{239,1,2,3},{0,0,0,0}}}, {active,false}]),
  {ok, S} = gen_udp:open(0, [{reuseaddr,true},binary]),
  ok = gen_udp:send(S, {239,1,2,3}, 4321, <<"aaaaa">>),
  {ok, {_,_,<<"aaaaa">>}} = gen_udp:recv(R, 0, 200).
```
```
% ./local/issue/udp_binary.erl
escript: exception error: no match of right hand side value 
                 {ok,{{172,17,0,4},48327,"aaaaa"}}
  in function  udp_binary_erl__escript__1668__593128__858464__2:main/1 (./local/issue/udp_binary.erl, line 9)
```

It works if put the `binary` after the `add_membership` option.

**Expected behavior**
This script runs without errors.

**Affected versions**
25.1.2

**Additional context**
In version 24.3.4.6 `binary` works when placed before or after the `add_membership` option.

**Comment by bmk** [comment]
Found the problem. Testing now.

**Comment by door** [comment]
There is yet another issue with `add_membership` and gen_udp:open. It returns einval when using `ip` after `add_membership`, but works as expected when `ip` goes before `add_membership`:

os: linux

erlang 24.3.4:
```
1> gen_udp:open(4321, [{add_membership,{{239,1,2,3},{0,0,0,0}}},{ip,{239,1,2,3}}]).
{ok,#Port<0.5>}
```
erlang 25.1.2:
```
1> gen_udp:open(4321, [{add_membership,{{239,1,2,3},{0,0,0,0}}},{ip,{239,1,2,3}}]).
{error,einval}
2> gen_udp:open(4321, [{ip,{239,1,2,3}},{add_membership,{{239,1,2,3},{0,0,0,0}}}]).
{ok,#Port<0.6>}
```

**Comment by bmk** [comment]
Its the same fault. I cannot reproduce after my fix.
Here is a branch with my correction:
https://github.com/bmk/otp/tree/bmk/kernel/20221116/gen_udp_opts_processing/OTP-18323

**Comment by door** [comment]
Thanks for prompt response!

The problem with einval and add_membership,ip opts is gone, but the first problem with binary still present.

**Comment by bmk** [comment]
Hum, I only looked at the option processing, but maybe there is something else...
...back to the workbench...

**Comment by bmk** [comment]
There was *also* a bug in the inet-driver (processing the [add|drop]_membership options).
Branch updated.

**Comment by door** [comment]
Both errors are gone, thank you!

**Comment by bmk** [comment]
Planned to be released with the 25.1.3 patch (whenever that is).
